contentcuration/contentcuration/frontend/channelEdit/components/ContentNodeValidator.vue

```diff
@@ -43,7 +45,7 @@
           : '';
       },
       error() {
-        if (!this.node.complete) {
+        if (!isNodeComplete({ nodeDetails: this.node })) {
```

Comment:
For some nodes, this could evaluate as `false` even when the node would be complete. For example, if a node is an exercise, `isNodeComplete` also needs to receive its assessment items. Otherwise, it evaluates it as missing questions and marks the node as incomplete.

There are two solutions:

(1) Use logic [like here](https://github.com/learningequality/studio/blob/develop/contentcuration/contentcuration/frontend/channelEdit/components/edit/DetailsTabView.vue#L553)
(2) Revert this change

I would suggest going with (2) because we need to investigate the issue you reported anyway, and I am 100% not sure if it's a question of the logic around `isNodeComplete`. There might be something weird going on with async behavior.

Also, `ContentNodeValidator` is not responsible for updating the node. It only reads the information. The idea behind updating `complete` field with `isNodeComplete` is to update it on each node change that is supposed to trigger validation, so this should not be necessary after we locate what's the real problem.

After we merge this PR, I am going to investigate further and will also add comment to `isNodeComplete` for developers, so we don't forget to provide all the information. Hopefully, we'll move this to backend in the future.

Reply:
I am thinking that I'll refactor `isNodeComplete` to `isExerciseComplete`, `isTopicComplete` etc. so it's more clear.

Reply:
It may not be necessary to split it - but I would add some type checks and error messages if you call it with an exercise but `undefined` assessment items then it should let you know?

I do think that fixing the issue _where_ the node is supposed to get completed so that completed nodes are indeed `node.complete === true` when it's supposed to be - I dug around for where that happens but wherever I led made me think that it was being set `complete` on the backend.

If you can guide me to the part of the code where this is happening, I'll update this PR to fix that - then will revert this change.

Reply:
> It may not be necessary to split it - but I would add some type checks and error messages if you call it with an exercise but undefined assessment items then it should let you know?

Yes, I am going to add error messages in any case and regarding splitting -  I will play with it and see what's better readable.

> I dug around for where that happens but wherever I led made me think that it was being set complete on the backend.
 If you can guide me to the part of the code where this is happening, I'll update this PR to fix that - then will revert this change.

It's happening on node fields updates - right now it's [details tab](https://github.com/learningequality/studio/blob/develop/contentcuration/contentcuration/frontend/channelEdit/components/edit/DetailsTabView.vue#L561) and [questions tab](https://github.com/learningequality/studio/blob/develop/contentcuration/contentcuration/frontend/channelEdit/components/AssessmentTab/AssessmentTab.vue#L117).

Might be good to check whether it's still happening at first because `diffTracker` could also cause some async inconsistencies, and I noticed that there was a fix merged for that meanwhile.

Reply:
I still ran into the issue again but in the reverse direction where a node was definitely complete, but showed `complete: false` in the Vuex state tree.

There is something just __slightly__ off somewhere I think. 

I was thinking maybe we should trigger the update with an empty payload here:

https://github.com/learningequality/studio/blob/6859e0476ec2cf08afccea0ded1581eca21250f9/contentcuration/contentcuration/frontend/channelEdit/components/edit/DetailsTabView.vue#L599

So that we are also always checking if the node is valid before loading the tab. Unfortunately, without a reliably replication of the bug I'm seeing I cannot say for sure if this is even a fix for it but I got the impression that the components were working with outdated data. Reliably replication would go a long way to helping figure this problem but no luck :(

Reply:
Thank you. Yes, it is and will be possible to retrieve data that don't have `complete` set because we lack backend `complete` validation at this point. It makes sense to run this check even before we update some fields of a node as you suggest. DetailsTabView might not be the only entry point though. I am going to check if it would be possible to do it on edit modal load.